# Let's see just how accurate this model is by using some test data!

In [1]:
from locale import normalize
import numpy as np
from keras.models import load_model
from keras.preprocessing import image
import keras
import cv2

In [2]:
TEST_DIR =  "C:\Build\SnakeImages\\test"

test_datagen = image.ImageDataGenerator( rescale= 1. / 255)
test_data = test_datagen.flow_from_directory(directory=TEST_DIR , target_size=(224,224) , batch_size=32 , class_mode='binary')

Found 200 images belonging to 3 classes.


In [3]:
# lets print the classes :

print("test_data.class_indices: ", test_data.class_indices)

test_data.class_indices:  {'NoSnake': 0, 'Snake': 1, 'test': 2}


In [4]:
#load the saved model :

model = load_model('C:\Build\SnakeImages\MyBestModel.h5')

#print(model.summary() )

acc = model.evaluate(x=test_data)[1]

print(acc)

7/7 [==============================] - 33s 4s/step - loss: 0.4315 - accuracy: 0.8100
0.8100000023841858


In [7]:
# load an image from the test folder 
imagePath = r"C:\Build\SnakeImages\test\Snake\image14_1.png"  # having some trouble here

img = image.load_img(imagePath,target_size=(224,224))
i = image.img_to_array(img) # convert to array
i = i / 255 # -> normalize to our model
print(i.shape)

input_arr = np.array([i]) # add another dimention 
print(input_arr.shape)

(224, 224, 3)
(1, 224, 224, 3)


In [8]:
# run the prediction
predictions = model.predict(input_arr)[0][0]
print(predictions)

1/1 [==============================] - 2s 2s/step
0.9956823


In [9]:
# since it is binary if the result is close to 0 there is a snake , and if it close to 1 there is no snake
result = round(predictions)
if result == 0 :
    text = 'Has a snake'
else :
    text = "No snake"


print(text)

No snake


In [10]:
imgResult = cv2.imread(imagePath)
font = cv2.FONT_HERSHEY_COMPLEX
cv2.putText(imgResult, text, (0,20), font, 0.8 , (255,0,0),2 )
cv2.imshow('img', imgResult)
cv2.waitKey(0)
cv2.imwrite("C:\Build\SnakeImages\predictImage.jpg",imgResult)

True

# Got an accuracy score of 0.8100, so either that is not so good or there is some confusion/noise within the data or model?
#### Ideal is above 95% based on a Google search